syntax = "proto3";

package project_data;

// Protobuf 消息定义...

// 前景掩码消息
message ForegroundMask {
  bytes mask_data = 1;  // 前景掩码的字节数据
  int32 width = 2;      // 掩码图像的宽度
  int32 height = 3;     // 掩码图像的高度
}

// 卡尔曼滤波状态
message State {
  float x = 1;
  float y = 2;
  float vx = 3;
  float vy = 4;
}

message StateUpdate {
  State state = 1;
}

// 对象检测结果
message Detection {
  float x1 = 1;
  float y1 = 2;
  float x2 = 3;
  float y2 = 4;
  float confidence = 5;
  string label = 6;
}

message DetectionResult {
  repeated Detection detections = 1;
}

// 图像数据
message Image {
  int32 width = 1;  // 图像宽度
  int32 height = 2;  // 图像高度
  string format = 3;  // 图像格式，如"JPEG"、"PNG"
  bytes data = 4;  // 图像数据，二进制格式
}

// gRPC 服务定义

service VideoProcessingService {
  rpc ProcessVideo(stream Image) returns (stream DetectionResult);
}

service ObjectDetectionService {
  rpc DetectObject(Image) returns (DetectionResult);
}

service BackgroundSubtractionService {
  rpc ApplyBackgroundSubtraction(Image) returns (ForegroundMask);
}

service KalmanFilterService {
  rpc UpdateState(DetectionResult) returns (StateUpdate);
}
